### PR TITLE
fix: plugin types compatibility issue with ctx.has

### LINKED ISSUE #263: bug: Type 'MyContext' does not satisfy the constraint 'Context'. Types of property 'has' are incompatible.
Hello!

I get the error `Type 'MyContext' does not satisfy the constraint 'Context'. Types of property 'has' are incompatible. ` when I try to create a new Bot. The problem only happens with version 1.10 of Grammy and not with the 1.9.2 version.

```ts
type MyContext = HydrateFlavor<Context> & SessionFlavor<SessionData>
type MyApi = HydrateApiFlavor<Api>

export const bot = new Bot<MyContext, MyApi>(`${process.env.BOT_TOKEN}`)
```

photo_2022-08-13_12-38-37

Thanks.

### PR DESCRIPTION
**Pull request by Loskir**
Fixes #263

### CODE REVIEW

**Comment by Loskir** [comment]
This PR fixes the issue that resulted in typescript error being thrown when extending `C extends Context` `C & {something: 1}` (as discussed [in Telegram](https://t.me/grammyjs/70601))

However, it features additive implementation of `NarrowMatch` that was [proven](https://t.me/grammyjs/70653) to work not as expected.

I think that this caveat is minor compared to the original issue. If we are unable to come up with a better alternative, this should be merged until we either find a better solution or split `match` into two properties in 2.0.

The original problem is very serious as it breaks numerous plugins

File: src/filter.ts [hunk 1]
```diff
@@ -433,6 +433,10 @@ type Combine<U, K extends string> = U extends unknown
     ? U & Partial<Record<Exclude<K, keyof U>, undefined>>
     : never;
 
+export type FilterE<Q extends FilterQuery> = PerformQueryE<
```
Comment (KnorpelSenf) [inline src/filter.ts:436]: This is a great PQ, I'd just love to have a better suffix than `E`. What does it stand for, extension maybe?

Comment (Loskir) [reply src/filter.ts:436]: Yes. I agree :D

Comment (KnorpelSenf) [reply src/filter.ts:436]: Let me throw a bunch of random names at you:
- FilterExtension
- FilterCore
- FilterAddition
- FilterCasing
- FilterShell
- FilterCover
- FilterSide
- FilterWing
- FilterAugmentation
- FilterEnhancement
- FilterAppendix
- FilterAttachment
- FilterAddendum

I think I prefer `FilterCore`. Which one do you like best?

Comment (Loskir) [reply src/filter.ts:436]: `FilterCore`, too

Comment (KnorpelSenf) [reply src/filter.ts:436]: Alright, let's rename then. Can you take care of this?

Comment (Loskir) [reply src/filter.ts:436]: What about exports and typedoc comments?

Comment (KnorpelSenf) [reply src/filter.ts]: I don't really think we need to export the `*Core` types. I see no reason to do this, as userland code would likely always want to use the full types. If there's ever demand for this, we can still export it in the future.

JSDoc comments are only necessary for exported stuff. Internal stuff may or may not have it, I'm mostly only doing this when things are hard to understand without comments.

Comment (Loskir) [reply src/filter.ts]: Great. Then it's done

**Comment by KnorpelSenf** [comment]
@EdJoPaTo should we wait for your review?

**Review by EdJoPaTo — APPROVED** [review]

File: src/context.ts [hunk 2]
```diff
@@ -2124,10 +2149,7 @@ export type InlineQueryContext<C extends Context> = Filter<
  */
 export type ChatTypeContext<C extends Context, T extends Chat["type"]> =
     & C
-    & Record<"update", ChatTypeUpdate<T>> // ctx.update
-    & ChatType<T> // ctx.chat
-    & ChatTypeRecord<"msg", T> // ctx.msg
-    & AliasProps<ChatTypeUpdate<T>>; // ctx.message etc
+    & ChatTypeContextCore<T>; // ctx.message etc
```
Comment (Loskir) [inline src/context.ts:2152]:
```suggestion
    & ChatTypeContextCore<T>;
```

File: src/context.ts [hunk 3]
```diff
@@ -2112,6 +2131,12 @@ export type InlineQueryContext<C extends Context> = Filter<
     C,
     "inline_query"
 >;
+
+type ChatTypeContextCore<T extends Chat["type"]> =
+    & Record<"update", ChatTypeUpdate<T>> // ctx.update
+    & ChatType<T> // ctx.chat
+    & ChatTypeRecord<"msg", T> // ctx.msg
+    & AliasProps<ChatTypeUpdate<T>>;
```
Comment (Loskir) [inline src/context.ts:2139]:
```suggestion
    & AliasProps<ChatTypeUpdate<T>>; // ctx.message etc
```

**Review by KnorpelSenf — APPROVED** [review]
Thank you so much!

This fix will be shipped alongside Bot API 6.2 support.

**Comment by KnorpelSenf** [comment]
@Loskir please consider signing your commits.

I'll merge anyway now, but this would be nice to have for future contributions :)

**Pull request merged**